# HG changeset patch
# User Sergey Mironov <grrwlf@gmail.com>
# Date 1387191221 -14400
#      Mon Dec 16 14:53:41 2013 +0400
# Node ID 2c19b7865d72ea34fae3e96c09fffba233fad873
# Parent  fda9d5af69e7443311aabf32bd444ab41107a74a
[Draft] loopback request for internal http server

diff --git a/include/urweb/urweb_cpp.h b/include/urweb/urweb_cpp.h
--- a/include/urweb/urweb_cpp.h
+++ b/include/urweb/urweb_cpp.h
@@ -21,6 +21,7 @@
 failure_kind uw_initialize(struct uw_context *);
 
 struct uw_context * uw_init(int id, void *logger_data, uw_logger log_debug);
+int uw_context_id(uw_context ctx);
 void uw_close(struct uw_context *);
 int uw_set_app(struct uw_context *, uw_app*);
 uw_app *uw_get_app(struct uw_context *);
diff --git a/src/c/http.c b/src/c/http.c
--- a/src/c/http.c
+++ b/src/c/http.c
@@ -5,12 +5,14 @@
 #include <stdlib.h>
 #include <sys/types.h>
 #include <sys/socket.h>
+#include <sys/stat.h>
 #include <netinet/in.h>
 #include <netinet/tcp.h>
 #include <arpa/inet.h>
 #include <unistd.h>
 #include <signal.h>
 #include <stdarg.h>
+#include <fcntl.h>
 
 #include <pthread.h>
 
@@ -70,6 +72,42 @@
   }
 }
 
+int mimic_send(int sock, const void *buf, ssize_t len)
+{
+  printf("mimic send: %d\n", (int) len);
+  return 0;
+}
+
+void mimic_request(uw_context caller_ctx, const char* c_url)
+{
+  printf("mimic uw_request: starting\n");
+
+  int me = uw_context_id(caller_ctx);
+  uw_context ctx = uw_request_new_context(me, &uw_application, NULL, log_error, log_debug);
+  uw_request_context rc = uw_new_request_context();
+
+  char path[256];
+  strcpy(path,c_url);
+
+  char *method = "GET";
+  char *query_string = "";
+  char *body = NULL;
+  size_t body_len = 0;
+
+  request_result rr;
+
+  int nullfd = open("/dev/null", O_WRONLY);
+
+
+  rr = uw_request(rc, ctx, method, path, query_string, body, body_len,
+                  on_success, on_failure,
+                  NULL, log_error, log_debug,
+                  nullfd, mimic_send, close);
+
+  printf("mimic uw_request: ret %d\n", rr);
+
+}
+
 static void *worker(void *data) {
   int me = *(int *)data;
   uw_context ctx = uw_request_new_context(me, &uw_application, NULL, log_error, log_debug);
diff --git a/src/c/urweb.c b/src/c/urweb.c
--- a/src/c/urweb.c
+++ b/src/c/urweb.c
@@ -481,6 +481,8 @@
 size_t uw_heap_max = SIZE_MAX;
 size_t uw_script_max = SIZE_MAX;
 
+int uw_context_id(uw_context ctx) { return ctx->id; }
+
 uw_context uw_init(int id, void *logger_data, uw_logger log_debug) {
   uw_context ctx = malloc(sizeof(struct uw_context));
 
